x/sophon/keeper/msg_server_credits.go

```diff
@@ -88,6 +89,145 @@ func (m msgServer) SettleCredits(goCtx context.Context, msg *types.MsgSettleCred
 }
 
-func (m msgServer) SubmitReports(_ context.Context, _ *types.MsgSubmitReports) (*types.MsgSubmitReportsResponse, error) {
-	panic("not implemented")
+func (m msgServer) SubmitReports(goCtx context.Context, msg *types.MsgSubmitReports) (*types.MsgSubmitReportsResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	sophonInfo, err := m.GetSophonInfo(ctx, msg.SophonPublicKey)
+	if err != nil {
+		if errors.Is(err, collections.ErrNotFound) {
+			return nil, sdkerrors.ErrNotFound.Wrapf("sophon not found for %s", msg.SophonPublicKey)
+		}
+		return nil, err
+	}
+
+	sophonPubKeyHex := hex.EncodeToString(msg.SophonPublicKey)
+
+	if sophonInfo.Address != msg.Address {
+		return nil, sdkerrors.ErrorInvalidSigner.Wrapf("invalid address; received %s, required %s", msg.Address, sophonInfo.Address)
+	}
+
+	events := sdk.Events{}
+
+	dataProxyCredits := make(map[string]math.Int)
+	totalUsedCredits := math.NewInt(0)
+
+	// There can be at most one report per user.
+	for _, report := range msg.Reports {
+		sophonUser, err := m.GetSophonUser(ctx, sophonInfo.Id, report.UserId)
+		if err != nil {
+			if errors.Is(err, collections.ErrNotFound) {
+				return nil, sdkerrors.ErrNotFound.Wrapf("user not found for %s", report.UserId)
+			}
+			return nil, err
+		}
+
+		if sophonUser.Credits.LT(report.UsedCredits) {
+			return nil, types.ErrInsufficientCredits.Wrapf("unable to submit report for user %s; requested %s, available %s", report.UserId, report.UsedCredits.String(), sophonUser.Credits.String())
+		}
+
+		// UsedCredits includes the data proxy credits so we don't need to check those separately.
+		sophonUser.Credits = sophonUser.Credits.Sub(report.UsedCredits)
+		totalUsedCredits = totalUsedCredits.Add(report.UsedCredits)
+
+		err = m.SetSophonUser(ctx, sophonInfo.Id, report.UserId, sophonUser)
+		if err != nil {
+			return nil, err
+		}
+
+		events = append(events,
+			sdk.NewEvent(
+				types.EventTypeUseUserCredits,
+				sdk.NewAttribute(types.AttributeSophonID, strconv.FormatUint(sophonInfo.Id, 10)),
+				sdk.NewAttribute(types.AttributeSophonPubKey, sophonPubKeyHex),
+				sdk.NewAttribute(types.AttributeUserID, report.UserId),
+				sdk.NewAttribute(types.AttributeCredits, report.UsedCredits.String()),
+			),
+			createUserEvent(sophonPubKeyHex, sophonInfo.Id, sophonUser),
+		)
+
+		for _, dataProxyReport := range report.DataProxyReports {
+			if dataProxyCredits[dataProxyReport.DataProxyPubKey].IsNil() {
+				dataProxyCredits[dataProxyReport.DataProxyPubKey] = math.NewInt(0)
+			}
+
+			dataProxyCredits[dataProxyReport.DataProxyPubKey] = dataProxyCredits[dataProxyReport.DataProxyPubKey].Add(dataProxyReport.Price.Mul(math.NewIntFromUint64(dataProxyReport.Amount)))
+		}
+
+	}
+
+	denom, err := m.stakingKeeper.BondDenom(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	totalDataProxyCredits := math.NewInt(0)
+	// Transfer the data proxy credits to the data proxy.
+	for dataProxyPubKey, amount := range dataProxyCredits {
```

Comment:
Since map iteration is not deterministic, I think we would have to just pay the data proxies in the earlier loop without constructing `dataProxyCredits`. Or `dataProxyCredits` could be a slice that we sort before making the payouts.